### DISCUSSION TITLE
Best practice to use WebSocket

### DISCUSSION THREAD

**Discussion by cflorion** [hummingbird-project/hummingbird#79] [Q&A] [answered]
Hi,

I am asking my question here because there is no Discussions tabs on https://github.com/hummingbird-project/hummingbird-websocket.

This is the first time I use Websocket via Swift-NIO and I am a bit lost.
I usually use https://github.com/Building42/Telegraph. 
Telegraph keeps an array of the connected webSockets so when I need to send then data, I loop over the array of connected webSockets and send them data individually.

With Hummingbird I am not sure how I should implement this behaviour.

``` 
app.ws.addUpgrade()
app.ws.add(middleware: HBLogRequestsMiddleware(.info))
app.ws.on("/ws") { [weak self] req, ws in
    ws.initiateAutoPing(interval: .seconds(60))
    ws.onClose { [weak self] (ws) in
    }
}
```

I wanted to create my own array of connected webSockets. When `app.ws.on("/ws")` is called, I add `ws` to my array. When `ws.onClose {...}` is called I remove `ws` from my array. 
But the webSocket object is not Hashable and have not unique identifier so I don't know how to remove it from the array.

I may have mistunderstood how webSockets work with NIO. Does someone have already done this ?
Thanks

**Comment by adam-fowler** [comment, marked as answer]
Yes the websocket code doesn't have any support for storing connections. I have left that up to you. The websocket chat example in hummingbird-examples has a basic connection manager https://github.com/hummingbird-project/hummingbird-examples/blob/main/websocket-chat/Sources/App/ConnectionManager.swift

**Reply by cflorion** [reply]
Makes sense! Thanks
